### PR TITLE
Add compiler-rt functions for f80

### PR DESCRIPTION
**Pull request by Vexu**
No tests but I have fuzzed them against hw a bit, might not work correctly for big-endian, and forcing f80 to be 16 bytes on x86 also seems to have some issues.

### CODE REVIEW

File: lib/std/math.zig [hunk 1]
```diff
@@ -43,11 +43,13 @@ pub const f128_max = @bitCast(f128, @as(u128, 0x7FFEFFFFFFFFFFFFFFFFFFFFFFFFFFFF
...
 pub const f128_toint = 1.0 / f128_epsilon;
 
-const F80Repr = if (@import("builtin").cpu.arch.endian() == .Little) extern struct {
+pub const F80Repr = if (@import("builtin").cpu.arch.endian() == .Little) extern struct {
     fraction: u64,
     exp: u16,
+    _pad: u32 = undefined,
 } else extern struct {
     exp: u16,
+    _pad: u32 = undefined,
```
[the remark this answers is not in the document]
Comment (matu3ba) [reply lib/std/math.zig:52]: It would be nice, if somewhere the device specific behavior could be linked.

**Review by matu3ba — CHANGES_REQUESTED** [review]
structure looks good, but I havent looked into the logical details of impl yet as this is still marked as draft.

mostly nits on some missing explanations and mask stuff (easy to introduce typos)

**Comment by matu3ba** [comment]
I think you can test this in a more structured manner with [the table from wikipedia](https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format). For the trunc stuff the musl libc tests are usually a great source, but in this case you could just use the test cases for Softfloat: http://www.jhauser.us/arithmetic/TestFloat.html

**Comment by Vexu** [comment]
> What's the merge checklist?

For little-endian non-i386 I just need to double check the `normalize_f80` function and address matu3ba's review.
The i386 issues are caused by stage1 using `LLVMABISizeOfType`/`LLVMABIAlignmentOfType` instead of `type->abi_size`/`type->abi_align` which causes mismatches in `get_llvm_type` when `f80` is used in a struct.

For big-endian rendering literals is still broken but that shouldn't block this.

And it wouldn't hurt to add some tests, I'll try generating some based on x86 hw implementation.

File: lib/std/special/compiler_rt/compareXf2.zig [hunk 2]
```diff
@@ -144,6 +144,69 @@ pub fn __gtdf2(a: f64, b: f64) callconv(.C) i32 {
...
+pub fn __eqxf2(a: f80, b: f80) callconv(.C) i32 {
+    return __lexf2(a, b);
+}
+
+pub fn __ltxf2(a: f80, b: f80) callconv(.C) i32 {
+    return __lexf2(a, b);
+}
+
+pub fn __nexf2(a: f80, b: f80) callconv(.C) i32 {
+    return __lexf2(a, b);
```
[the remark this answers is not in the document]
Comment (Vexu) [reply lib/std/special/compiler_rt/compareXf2.zig:203]: This is just to match the other `__neXf2` functions, the caller is supposed to compare the result to `0`.

File: lib/std/special/compiler_rt/compareXf2.zig [hunk 3]
```diff
@@ -144,6 +144,69 @@ pub fn __gtdf2(a: f64, b: f64) callconv(.C) i32 {
...
+}
+
+pub fn __gexf2(a: f80, b: f80) callconv(.C) i32 {
+    @setRuntimeSafety(builtin.is_test);
+    const float = cmp_f80(GE, a, b);
+    return @bitCast(i32, float);
+}
+
+pub fn __eqxf2(a: f80, b: f80) callconv(.C) i32 {
+    return __lexf2(a, b);
```
[the remark this answers is not in the document]
Comment (Vexu) [reply lib/std/special/compiler_rt/compareXf2.zig:195]: Same as the other.

File: lib/std/special/compiler_rt/compareXf2.zig [hunk 4]
```diff
@@ -144,6 +144,69 @@ pub fn __gtdf2(a: f64, b: f64) callconv(.C) i32 {
     return __gedf2(a, b);
 }
 
+// Comparison between f80
+
+pub inline fn cmp_f80(comptime RT: type, a: f80, b: f80) RT {
+    const a_rep = @ptrCast(*const std.math.F80Repr, &a).*;
+    const b_rep = @ptrCast(*const std.math.F80Repr, &b).*;
+
+    // If either a or b is NaN, they are unordered.
```
[the remark this answers is not in the document]
Comment (Vexu) [reply lib/std/special/compiler_rt/compareXf2.zig:153]: It is referencing NaN comparisons which are always unordered.
```
NaN != NaN == true
NaN == NaN == false
NaN > NaN == false
NaN >= NaN == false
NaN < NaN == false
NaN <= NaN == false
```

File: lib/std/special/compiler_rt/compareXf2.zig [hunk 5]
```diff
@@ -144,6 +144,69 @@ pub fn __gtdf2(a: f64, b: f64) callconv(.C) i32 {
...
+    if ((a_rep.exp & 0x7FFF == 0x7FFF and a_rep.fraction ^ 0x8000000000000000 != 0) or
+        (b_rep.exp & 0x7FFF == 0x7FFF and b_rep.fraction ^ 0x8000000000000000 != 0))
+        return RT.Unordered;
+
+    // If a and b are both zeros, they are equal.
+    if ((a_rep.fraction | b_rep.fraction) | ((a_rep.exp | b_rep.exp) & 0x7FFF) == 0)
+        return .Equal;
+
+    if (@boolToInt(a_rep.exp == b_rep.exp) & @boolToInt(a_rep.fraction == b_rep.fraction) != 0) {
```
[the remark this answers is not in the document]
Comment (Vexu) [reply lib/std/special/compiler_rt/compareXf2.zig:162]: This function is supposed to match hardware which compares bit for bit equality.

File: lib/std/special/compiler_rt/trunc_f80.zig [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        // The exponent of a is within the range of normal numbers in the
+        // destination format.  We can convert by simply right-shifting with
+        // rounding and adjusting the exponent.
+        abs_result = @as(dst_rep_t, a_rep.exp) << dst_sig_bits;
+        abs_result |= @truncate(dst_rep_t, a_rep.fraction >> (src_sig_bits - dst_sig_bits));
+        abs_result -%= @as(dst_rep_t, src_exp_bias - dst_exp_bias) << dst_sig_bits;
+
+        const round_bits = a_rep.fraction & round_mask;
+        if (round_bits > halfway) {
+            // Round to nearest
```
[the remark this answers is not in the document]
Comment (Vexu) [reply lib/std/special/compiler_rt/trunc_f80.zig:61]: Not sure, copied it from the `truncXfYf2` function.

**Pull request merged**